src/engine.ts

```diff
@@ -644,6 +720,68 @@ class StyleElementController extends NodeController<HTMLStyleElement> {
   }
 }
 
+class ShadowRootController extends NodeController<ShadowRoot> {
+  private controller: AbortController | null = null;
+  private mo: MutationObserver;
+  private host: HTMLElement;
+
+  constructor(node: ShadowRoot, mo: MutationObserver) {
+    super(node);
+    this.mo = mo;
+    this.host = node.host as HTMLElement;
+  }
+
+  connected(): void {
+    this.mo?.observe(this.node, {
+      childList: true,
+      subtree: true,
+      attributes: true,
+      attributeOldValue: true,
+    });
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    (this.host as any)[SHADOW_SYMBOL] = this.node;
```

Comment:
I think we should probably just set this in `attachShadow`.

Reply:
great call out here, that definitely makes more sense.